frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -46,185 +45,217 @@
+            LOGGER.logDebug("NodeGpuType does not match: Node: [%s] Request NodeGpuType: [%s], Available NodeGpuType: [%s]",
+                nodeHost, requestNodeGpuType, availableNodeGpuType);
+            filteredNodes.remove(i - 1);
             continue;
           }
-        } else {
-          LOGGER.logWarning(logPrefix +
-                  "Configured Nodes is not found in ClusterConfiguration: Ignore Request NodeGpuType: [%s]",
-              requestNodeGpuType);
         }
       }
+    }
+  }
 
-      // Test Resource
-      ResourceDescriptor availableResource = node.getAvailableResource();
-      if (!ResourceDescriptor.fitsIn(requestResource, availableResource)) {
-        LOGGER.logDebug(rejectedLogPrefix +
-                "Resource does not fit in: Request Resource: [%s], Available Resource: [%s]",
-            requestResource, availableResource);
-        continue;
+  private void filterNodesForNonGpuJob(int jobTotalRequestGpu) {
+    if (jobTotalRequestGpu == 0) {
+      for (int i = filteredNodes.size(); i > 0; i--) {
+        Node node = allNodes.get(filteredNodes.get(i - 1));
+        ResourceDescriptor totalResource = node.getTotalResource();
+        if (totalResource.getGpuNumber() > 0) {
+          LOGGER.logDebug("skip nodes with Gpu resource for 0-gpu job: Node [%s], Node total resource: [%s]",
+              node.getHost(), totalResource);
+          filteredNodes.remove(i - 1);
+        }
       }
+    }
+  }
 
-      // Test Optimized Resource
-      ResourceDescriptor optimizedRequestResource = YamlUtils.deepCopy(requestResource, ResourceDescriptor.class);
-      if (requestResource.getGpuAttribute() == 0) {
-        // If GpuAttribute is not explicitly specified, we select an optimal GpuAttribute according to 
-        // the current status of the node instead of let RM to select a random GpuAttribute.
-        optimizedRequestResource.setGpuAttribute(selectCandidateGpuAttribute(node, requestResource.getGpuNumber()));
-        if (!ResourceDescriptor.fitsIn(optimizedRequestResource, availableResource)) {
-          LOGGER.logDebug(rejectedLogPrefix +
-                  "Resource does not fit in: Optimized Request Resource: [%s], Available Resource: [%s]",
-              optimizedRequestResource, availableResource);
-          continue;
+  private void filterNodesByResource(ResourceDescriptor requestResource, Boolean skipLocalTriedResource) {
+    if (requestResource != null) {
+      for (int i = filteredNodes.size(); i > 0; i--) {
+        Node node = allNodes.get(filteredNodes.get(i - 1));
+        ResourceDescriptor availableResource = node.getAvailableResource();
+        if (skipLocalTriedResource && localTriedResource.containsKey(node.getHost())) {
+          LOGGER.logDebug("Skip local tried resources: [%s] on Node : [%s]", localTriedResource.get(node.getHost()), node.getHost());
+          ResourceDescriptor.subtractFrom(availableResource, localTriedResource.get(node.getHost()));
+        }
```

Comment:
You should not change the node.getAvailableResource() here. Besides, the node.getAvailableResource() is just a temp variable, change it has nothing impacts.

Reply:
the availableResource is already excluded the requested resource. you may double excluded it.

---
In reply to: [174701119](https://github.com/Microsoft/pai/pull/318#discussion_r174701119) [](ancestors = 174701119)